example/gmt/sphere_preprocessing.c

```diff
@@ -0,0 +1,525 @@
+{
+  if (fabs (xyz[0] + 0.5) < SC_EPS)
+    return 2;
+  if (fabs (xyz[0] - 0.5) < SC_EPS)
+    return 5;
+  if (fabs (xyz[1] + 0.5) < SC_EPS)
+    return 4;
+  if (fabs (xyz[1] - 0.5) < SC_EPS)
+    return 1;
+  if (fabs (xyz[2] + 0.5) < SC_EPS)
+    return 0;
+  if (fabs (xyz[2] - 0.5) < SC_EPS)
+    return 3;
+  return -1;                    /* This should not happen */
+}
+
+/**
+ * coordinate transformation from the surface of cube [-0.5,0.5]^3
+ * to tree-local coordinates. This is the inverse of 
+ * p4est_geometry_cubed_X
+ *
+ * \param[in]  xyz  cartesian coordinates in physical space
+ * \param[in]  which_tree face of cube
+ * \param[out] rst  coordinates in AMR space : [0,1]^3
+ *
+ */
+static void
+p4est_geometry_cubed_Y (const double xyz[3], double rst[3],
+                        p4est_topidx_t which_tree)
+{
+  rst[2] = 0.0;
+
+  /* align center with origin */
+  switch (which_tree) {
+  case 0:
+    rst[0] = xyz[1] + 0.5;
+    rst[1] = xyz[0] + 0.5;
+    break;
+  case 1:
+    rst[0] = xyz[2] + 0.5;
+    rst[1] = xyz[0] + 0.5;
+    break;
+  case 2:
+    rst[0] = xyz[2] + 0.5;
+    rst[1] = xyz[1] + 0.5;
+    break;
+  case 3:
+    rst[0] = xyz[0] + 0.5;
+    rst[1] = xyz[1] + 0.5;
+    break;
+  case 4:
+    rst[0] = xyz[0] + 0.5;
+    rst[1] = xyz[2] + 0.5;
+    break;
+  case 5:
+    rst[0] = xyz[1] + 0.5;
+    rst[1] = xyz[2] + 0.5;
+    break;
+  default:
+    break;
+  }
+}
+
+/** Returns 1 if the cone spanned by v1 and v2 intersects the line segment
+ *  between p1 and p2. If an intersection is detected then p_intersect is
+ *  set to the computed intersection point.
+ * 
+ *  We assume that v1 and v2 are not colinear. 
+ */
+static int
+cone_line_intersection (const double v1[3], const double v2[3],
+                        const double p1[3], const double p2[3],
+                        double p_intersect[3])
+{
+  /** We solve the matrix equation (v1, v2, p1-p2) x = p1 by inverting 
+   * (v1, v2, p1-p2) */
+  double              A[3][3];  /* Matrix we are inverting */
+  double              cofactor[3][3];   /* Cofactor matrix */
+  double              det_A, det_A_inv;
+  double              x[3];     /* Solving for x */
+
+  A[0][0] = v1[0];
+  A[1][0] = v1[1];
+  A[2][0] = v1[2];
+  A[0][1] = v2[0];
+  A[1][1] = v2[1];
+  A[2][1] = v2[2];
+  A[0][2] = p1[0] - p2[0];
+  A[1][2] = p1[1] - p2[1];
+  A[2][2] = p1[2] - p2[2];
+
+  /* Compute minors */
+  for (int r = 0; r < 3; r++) {
+    for (int c = 0; c < 3; c++) {
+      cofactor[r][c] =
+        A[(r + 1) % 3][(c + 1) % 3] * A[(r + 2) % 3][(c + 2) % 3] -
+        A[(r + 1) % 3][(c + 2) % 3] * A[(r + 2) % 3][(c + 1) % 3];
+    }
+  }
+
+  /* Compute the determinant by Laplace expansion along the first column */
+  det_A = 0;
+  for (int r = 0; r < 3; r++) {
+    det_A += A[r][0] * cofactor[r][0];
+  }
+
+  /** If the determinant is zero then the line segment is parallel to the 
+   * cone. We count this case as not intersecting.
+  */
+  if (det_A < SC_EPS) {
```

Comment:
Sometimes SC_1000_EPS is more robust. We need a well defined plan B in that case.

Reply:
I switched to SC_1000_EPS. It didn't make any difference in my real dataset anyway.